File: benchmarkDS/ourMethod/OPENAI_gpt-5-2025-08-07/Lse126.java
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;

public class Lse126 {
    private static final int MAX_USERNAME_CP = 256;

    // Constructs a safe HTML response for /hello using the given request body (username).
    public static String http_hello_page(String requestBody) {
        String username = requestBody == null ? "" : requestBody;
        // Trim surrounding whitespace
        username = username.trim();
        // Limit to MAX_USERNAME_CP code points to avoid excessive memory usage and DoS
        username = limitCodePoints(username, MAX_USERNAME_CP);
        // HTML-escape to prevent XSS
        String safeUsername = escapeHtml(username);

        StringBuilder outputbuf = new StringBuilder(256 + safeUsername.length());
        outputbuf.append("<!doctype html>")
                 .append("<html><head><meta charset=\"UTF-8\"><title>Hello</title></head><body>")
                 .append("<h1>Hello, ")
                 .append(safeUsername)
                 .append("!</h1>")
                 .append("</body></html>");
        return outputbuf.toString();
    }

    // Helper to limit a string to at most n Unicode code points without cutting surrogate pairs.
    private static String limitCodePoints(String s, int maxCp) {
        if (s == null) return "";
        int cpCount = s.codePointCount(0, s.length());
        if (cpCount <= maxCp) return s;
        int endIndex = s.offsetByCodePoints(0, maxCp);
        return s.substring(0, endIndex);
    }

    // Minimal HTML escaping
    private static String escapeHtml(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 16);
        for (int i = 0; i < s.length(); ) {
            int cp = s.codePointAt(i);
            i += Character.charCount(cp);
            switch (cp) {
                case '&': sb.append("&amp;"); break;
                case '<': sb.append("&lt;"); break;
                case '>': sb.append("&gt;"); break;
                case '"': sb.append("&quot;"); break;
                case '\'': sb.append("&#x27;"); break;
                default:
                    // Avoid control characters except common whitespace
                    if (cp >= 0x20 || cp == '\n' || cp == '\r' || cp == '\t') {
                        sb.appendCodePoint(cp);
                    } else {
                        sb.append(' ');
                    }
            }
        }
        return sb.toString();
    }

    public static void main(String[] args) throws IOException {
        String[] testRequests = new String[] {
            "Alice",
            "",
            "A".repeat(300), // will be truncated to 256 code points
            "<script>alert('x')</script>",
            "Élodie & 张伟"
        };

        for (int i = 0; i < testRequests.length; i++) {
            // Server reads request body (username)
            String username = testRequests[i];

            // Server constructs the response body (HTML)
            String outputbuf = http_hello_page(username);

            // Server writes the response body to the response stream
            ByteArrayOutputStream responseStream = new ByteArrayOutputStream();
            try {
                responseStream.write(outputbuf.getBytes(StandardCharsets.UTF_8));
                // Server closes the response stream
                responseStream.close();
            } finally {
                // Already closed by close(); ensure no resource leak
            }

            // For demonstration: print the response body
            System.out.println("=== Test Case " + (i + 1) + " ===");
            System.out.println(outputbuf);
            System.out.println();
        }
    }
}

File: benchmarkDS/ourMethod/OPENAI_gpt-5-2025-08-07/Lse126.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <limits>

static constexpr std::size_t MAX_USERNAME = 256;

// HTML escape to prevent XSS
std::string escape_html(const std::string& s) {
    std::string out;
    out.reserve(s.size() + 16);
    for (unsigned char ch : s) {
        switch (ch) {
            case '&': out += "&amp;"; break;
            case '<': out += "&lt;"; break;
            case '>': out += "&gt;"; break;
            case '"': out += "&quot;"; break;
            case '\'': out += "&#x27;"; break;
            default:
                // filter control chars except common whitespace
                if (ch >= 0x20 || ch == '\n' || ch == '\r' || ch == '\t') {
                    out.push_back(static_cast<char>(ch));
                } else {
                    out.push_back(' ');
                }
        }
    }
    return out;
}

// Constructs a safe HTML response for /hello using the given request body (username).
std::string http_hello_page(const std::string& request_body) {
    std::string username = request_body;
    // Trim leading/trailing whitespace
    auto lpos = username.find_first_not_of(" \t\r\n");
    auto rpos = username.find_last_not_of(" \t\r\n");
    if (lpos == std::string::npos) {
        username.clear();
    } else {
        username = username.substr(lpos, rpos - lpos + 1);
    }
    if (username.size() > MAX_USERNAME) {
        username.resize(MAX_USERNAME);
    }

    std::string safe = escape_html(username);
    std::ostringstream outputbuf;
    outputbuf << "<!doctype html>"
              << "<html><head><meta charset=\"UTF-8\"><title>Hello</title></head><body>"
              << "<h1>Hello, " << safe << "!</h1>"
              << "</body></html>";
    return outputbuf.str();
}

int main() {
    std::vector<std::string> tests = {
        "Alice",
        "",
        std::string(300, 'A'), // will be truncated to 256
        "<script>alert('x')</script>",
        "Élodie & 张伟"
    };

    for (std::size_t i = 0; i < tests.size(); ++i) {
        // Server reads request body (username)
        std::string username = tests[i];

        // Server constructs the response body (HTML)
        std::string outputbuf = http_hello_page(username);

        // Server writes the response body to the response stream
        std::ostringstream response_stream;
        response_stream << outputbuf;
        response_stream.flush();
        // "Close" the response stream by destroying it at end of scope

        // For demonstration: print the response body
        std::cout << "=== Test Case " << (i + 1) << " ===\n";
        std::cout << response_stream.str() << "\n\n";
    }
    return 0;
}